compiler/rustc_lint/src/redundant_sizedness_bound.rs

```diff
@@ -0,0 +1,240 @@
+use rustc_errors::Applicability;
+use rustc_hir::def_id::{DefId, DefIdMap};
+use rustc_hir::{
+    BoundPolarity, GenericBound, Generics, PolyTraitRef, TraitBoundModifiers, WherePredicateKind,
+};
+use rustc_middle::ty::{ClauseKind, PredicatePolarity};
+use rustc_session::{declare_lint, declare_lint_pass};
+use rustc_span::symbol::Ident;
+
+use crate::{LateContext, LateLintPass, LintContext};
+
+declare_lint! {
+    /// The `redundant_sizedness_bound` lint detects redundant sizedness bounds applied to type parameters that are already
+    /// otherwise implied.
+    ///
+    /// ### Example
+    ///
+    /// ```rust
+    /// #![feature(sized_hierarchy)]
+    /// use std::marker::MetaSized;
+    /// // `T` must be `Sized` due to the bound `Clone`, thus `?Sized` is redundant.
+    /// fn f<T: Clone + ?Sized>(t: &T) {}
+    /// // `T` is `Sized` due to `Clone` bound, thereby implying `MetaSized` and making the explicit `MetaSized` bound redundant.
+    /// fn g<T: MetaSized + Clone>(t: &T) {}
+    /// ```
+    ///
```

Comment:
This comment produces output in our user-facing documentation [here](https://doc.rust-lang.org/rustc/lints/listing/warn-by-default.html).  Generally we don't mention unstable features in this kind of documentation (as, e.g., it would make a reader feel like the person should know what `MetaSized` and its properties are).  Probably it's worth making this example self-standing without unstable features.